README.md
# fastqueue
Simple queue system based on FastAPI and PostgreSQL.

## Features

- Simple rest api.
- Message filtering support.
- Dead queue support.
- Redrive support (move messages between queues).
- Delay queues support.
- Prometheus metrics support.
- Simplicity, it does the minimum necessary, it will not have an authentication/permission scheme among other things.

## Quickstart

Let's start with the basic concepts, we have three main entities that we must know to start:

- Topic: A named resource to which messages are sent.
- Queue: A named resource representing a queue that receives messages from topics.
- Message: The data that a publisher sends to a topic is eventually delivered to queues.

### Environment variables

See https://github.com/allisson/fastqueue/blob/main/env.sample.default.

### Run the local PostgreSQL server

To run the server it is necessary to have a database available from PostgreSQL:

```bash
docker run --name postgres-fastqueue \
    --restart unless-stopped \
    -e POSTGRES_USER=fastqueue \
    -e POSTGRES_PASSWORD=fastqueue \
    -e POSTGRES_DB=fastqueue \
    -p 5432:5432 \
    -d postgres:14-alpine
```

### Run the database migration

The database migration is responsible to create the database schema.

```bash
docker run --rm \
    -e fastqueue_database_url='postgresql+psycopg2://fastqueue:fastqueue@localhost:5432/fastqueue' \
    --network="host" \
    quay.io/allisson/fastqueue db-migrate
```

### Run the worker

The worker is responsible for cleanup the messages from queues (remove expired messages and move to dead queue if configured).

```bash
docker run --name fastqueue-worker \
    --restart unless-stopped \
    -e fastqueue_database_url='postgresql+psycopg2://fastqueue:fastqueue@localhost:5432/fastqueue' \
    -e fastqueue_queue_cleanup_interval_seconds=60 \
    --network="host" \
    quay.io/allisson/fastqueue worker
```

### Run the server

The server is responsible to deliver the rest API.

```bash
docker run --name fastqueue-server \
    --restart unless-stopped \
    -e fastqueue_database_url='postgresql+psycopg2://fastqueue:fastqueue@localhost:5432/fastqueue' \
    -e fastqueue_server_port=8000 \
    -e fastqueue_server_num_workers=1 \
    --network="host" \
    quay.io/allisson/fastqueue server
```

You can access the api docs at http://localhost:8000/docs or http://localhost:8000/redoc.

### Create a new topic

```bash
curl -i -X 'POST' \
  'http://127.0.0.1:8000/topics' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "id": "events"
}'

HTTP/1.1 201 Created
date: Wed, 05 Oct 2022 21:48:24 GMT
server: uvicorn
content-length: 57
content-type: application/json

{
  "id":"events",
  "created_at":"2022-10-05T21:48:24.724341"
}
```

```bash
curl -i -X 'GET' \
  'http://127.0.0.1:8000/topics' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Wed, 05 Oct 2022 21:49:40 GMT
server: uvicorn
content-length: 68
content-type: application/json

{
  "data":[
    {
      "id":"events",
      "created_at":"2022-10-05T21:48:24.724341"
    }
  ]
}
```

### Create a new queue

```bash
curl -i -X 'POST' \
  'http://localhost:8000/queues' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "id": "all-events",
  "topic_id": "events",
  "ack_deadline_seconds": 30,
  "message_retention_seconds": 1209600
}'

HTTP/1.1 201 Created
date: Wed, 05 Oct 2022 21:51:43 GMT
server: uvicorn
content-length: 259
content-type: application/json

{
  "id":"all-events",
  "topic_id":"events",
  "dead_queue_id":null,
  "ack_deadline_seconds":30,
  "message_retention_seconds":1209600,
  "message_filters":null,
  "message_max_deliveries":null,
  "delivery_delay_seconds":null,
  "created_at":"2022-10-05T21:51:44.684743",
  "updated_at":"2022-10-05T21:51:44.684743"
}
```

### Create a new message

```bash
curl -i -X 'POST' \
  'http://localhost:8000/topics/events/messages' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "data": {"event_name": "event1", "success": true},
  "attributes": {"event_name": "event1"}
}'

HTTP/1.1 201 Created
date: Wed, 05 Oct 2022 21:54:07 GMT
server: uvicorn
content-length: 355
content-type: application/json

{
  "data":[
    {
      "id":"147d5f58-9dc5-4f69-8ab1-107a799fd731",
      "queue_id":"all-events",
      "data":{
        "event_name":"event1",
        "success":true
      },
      "attributes":{
        "event_name":"event1"
      },
      "delivery_attempts":0,
      "expired_at":"2022-10-19T21:54:08.010379",
      "scheduled_at":"2022-10-05T21:54:08.010379",
      "created_at":"2022-10-05T21:54:08.010379",
      "updated_at":"2022-10-05T21:54:08.010379"
    }
  ]
}
```

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/all-events/stats' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Wed, 05 Oct 2022 21:55:13 GMT
server: uvicorn
content-length: 70
content-type: application/json

{
  "num_undelivered_messages":1,
  "oldest_unacked_message_age_seconds":66
}
```

### Consume messages from the queue

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/all-events/messages' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Wed, 05 Oct 2022 21:56:07 GMT
server: uvicorn
content-length: 355
content-type: application/json

{
  "data":[
    {
      "id":"147d5f58-9dc5-4f69-8ab1-107a799fd731",
      "queue_id":"all-events",
      "data":{
        "success":true,
        "event_name":"event1"
      },
      "attributes":{
        "event_name":"event1"
      },
      "delivery_attempts":1,
      "expired_at":"2022-10-19T21:54:08.010379",
      "scheduled_at":"2022-10-05T21:56:37.272384",
      "created_at":"2022-10-05T21:54:08.010379",
      "updated_at":"2022-10-05T21:56:07.272384"
    }
  ]
}
```

```bash
curl -i -X 'PUT' \
  'http://localhost:8000/messages/147d5f58-9dc5-4f69-8ab1-107a799fd731/nack' \
  -H 'accept: application/json'

HTTP/1.1 204 No Content
date: Wed, 05 Oct 2022 21:57:17 GMT
server: uvicorn
content-type: application/json
```

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/all-events/messages' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Wed, 05 Oct 2022 21:57:53 GMT
server: uvicorn
content-length: 355
content-type: application/json

{
  "data":[
    {
      "id":"147d5f58-9dc5-4f69-8ab1-107a799fd731",
      "queue_id":"all-events",
      "data":{
        "success":true,
        "event_name":"event1"
      },
      "attributes":{
        "event_name":"event1"
      },
      "delivery_attempts":2,
      "expired_at":"2022-10-19T21:54:08.010379",
      "scheduled_at":"2022-10-05T21:58:24.560500",
      "created_at":"2022-10-05T21:54:08.010379",
      "updated_at":"2022-10-05T21:57:54.560500"
    }
  ]
}
```

```bash
curl -i -X 'PUT' \
  'http://localhost:8000/messages/147d5f58-9dc5-4f69-8ab1-107a799fd731/ack' \
  -H 'accept: application/json'

HTTP/1.1 204 No Content
date: Wed, 05 Oct 2022 21:58:56 GMT
server: uvicorn
content-type: application/json
```

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/all-events/messages' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Wed, 05 Oct 2022 21:59:26 GMT
server: uvicorn
content-length: 11
content-type: application/json

{
  "data":[]
}
```

## Message filtering

We can receive a subset of the messages in the queue using the message_filters field.

```bash
curl -i -X 'POST' \
  'http://localhost:8000/queues' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "id": "filtered-events",
  "topic_id": "events",
  "ack_deadline_seconds": 30,
  "message_retention_seconds": 1209600,
  "message_filters": {
    "event_name": ["event2"]
  }
}'

HTTP/1.1 201 Created
date: Wed, 05 Oct 2022 23:42:09 GMT
server: uvicorn
content-length: 285
content-type: application/json

{
  "id":"filtered-events",
  "topic_id":"events",
  "dead_queue_id":null,
  "ack_deadline_seconds":30,
  "message_retention_seconds":1209600,
  "message_filters":{
    "event_name":[
      "event2"
    ]
  },
  "message_max_deliveries":null,
  "delivery_delay_seconds":null,
  "created_at":"2022-10-05T23:42:10.322336",
  "updated_at":"2022-10-05T23:42:10.322336"
}
```

```bash
curl -i -X 'POST' \
  'http://localhost:8000/topics/events/messages' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "data": {"event_name": "event1", "success": true},
  "attributes": {"event_name": "event1"}
}'

HTTP/1.1 201 Created
date: Wed, 05 Oct 2022 23:43:33 GMT
server: uvicorn
content-length: 355
content-type: application/json

{
  "data":[
    {
      "id":"1dd422ad-7d20-4257-94c0-00fc1fbb8092",
      "queue_id":"all-events",
      "data":{
        "event_name":"event1",
        "success":true
      },
      "attributes":{
        "event_name":"event1"
      },
      "delivery_attempts":0,
      "expired_at":"2022-10-19T23:43:33.265369",
      "scheduled_at":"2022-10-05T23:43:33.265369",
      "created_at":"2022-10-05T23:43:33.265369",
      "updated_at":"2022-10-05T23:43:33.265369"
    }
  ]
}
```

```bash
curl -i -X 'POST' \
  'http://localhost:8000/topics/events/messages' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "data": {"event_name": "event2", "success": true},
  "attributes": {"event_name": "event2"}
}'

HTTP/1.1 201 Created
date: Wed, 05 Oct 2022 23:51:35 GMT
server: uvicorn
content-length: 705
content-type: application/json

{
  "data":[
    {
      "id":"8856ea21-665f-40a1-b576-33fa067c6a7a",
      "queue_id":"all-events",
      "data":{
        "event_name":"event2",
        "success":true
      },
      "attributes":{
        "event_name":"event2"
      },
      "delivery_attempts":0,
      "expired_at":"2022-10-19T23:51:36.175915",
      "scheduled_at":"2022-10-05T23:51:36.175915",
      "created_at":"2022-10-05T23:51:36.175915",
      "updated_at":"2022-10-05T23:51:36.175915"
    },
    {
      "id":"46f9b01c-2a1c-46d9-89a2-0beb2c331ae3",
      "queue_id":"filtered-events",
      "data":{
        "event_name":"event2",
        "success":true
      },
      "attributes":{
        "event_name":"event2"
      },
      "delivery_attempts":0,
      "expired_at":"2022-10-19T23:51:36.176050",
      "scheduled_at":"2022-10-05T23:51:36.176050",
      "created_at":"2022-10-05T23:51:36.176050",
      "updated_at":"2022-10-05T23:51:36.176050"
    }
  ]
}
```

## Dead queue support

The idea of the dead queue is to move messages that could not be processed to another queue, this can be done using the combination of dead_queue_id and message_max_deliveries fields.

```bash
curl -i -X 'POST' \
  'http://localhost:8000/queues' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "id": "all-events-dead",
  "ack_deadline_seconds": 30,
  "message_retention_seconds": 1209600
}'

HTTP/1.1 201 Created
date: Thu, 06 Oct 2022 00:33:03 GMT
server: uvicorn
content-length: 260
content-type: application/json

{
  "id":"all-events-dead",
  "topic_id":null,
  "dead_queue_id":null,
  "ack_deadline_seconds":30,
  "message_retention_seconds":1209600,
  "message_filters":null,
  "message_max_deliveries":null,
  "delivery_delay_seconds":null,
  "created_at":"2022-10-06T00:33:04.707829",
  "updated_at":"2022-10-06T00:33:04.707829"
}
```

```bash
curl -i -X 'PUT' \
  'http://localhost:8000/queues/all-events' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "topic_id": "events",
  "ack_deadline_seconds": 30,
  "message_retention_seconds": 1209600,
  "dead_queue_id": "all-events-dead",
  "message_max_deliveries": 2
}'

HTTP/1.1 200 OK
date: Thu, 06 Oct 2022 01:05:55 GMT
server: uvicorn
content-length: 269
content-type: application/json

{
  "id":"all-events",
  "topic_id":"events",
  "dead_queue_id":"all-events-dead",
  "ack_deadline_seconds":30,
  "message_retention_seconds":1209600,
  "message_filters":null,
  "message_max_deliveries":2,
  "delivery_delay_seconds":null,
  "created_at":"2022-10-05T21:51:44.684743",
  "updated_at":"2022-10-06T01:05:56.066023"
}
```

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/all-events/messages' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Thu, 06 Oct 2022 01:07:50 GMT
server: uvicorn
content-length: 700
content-type: application/json

{
  "data":[
    {
      "id":"1dd422ad-7d20-4257-94c0-00fc1fbb8092",
      "queue_id":"all-events",
      "data":{
        "success":true,
        "event_name":"event1"
      },
      "attributes":{
        "event_name":"event1"
      },
      "delivery_attempts":1,
      "expired_at":"2022-10-19T23:43:33.265369",
      "scheduled_at":"2022-10-06T01:08:21.408095",
      "created_at":"2022-10-05T23:43:33.265369",
      "updated_at":"2022-10-06T01:07:51.408095"
    },
    {
      "id":"8856ea21-665f-40a1-b576-33fa067c6a7a",
      "queue_id":"all-events",
      "data":{
        "success":true,
        "event_name":"event2"
      },
      "attributes":{
        "event_name":"event2"
      },
      "delivery_attempts":1,
      "expired_at":"2022-10-19T23:51:36.175915",
      "scheduled_at":"2022-10-06T01:08:21.408095",
      "created_at":"2022-10-05T23:51:36.175915",
      "updated_at":"2022-10-06T01:07:51.408095"
    }
  ]
}
```

```bash
curl -i -X 'PUT' \
  'http://localhost:8000/messages/1dd422ad-7d20-4257-94c0-00fc1fbb8092/nack' \
  -H 'accept: application/json'

HTTP/1.1 204 No Content
date: Thu, 06 Oct 2022 01:09:54 GMT
server: uvicorn
content-type: application/json
```

```bash
curl -i -X 'PUT' \
  'http://localhost:8000/messages/8856ea21-665f-40a1-b576-33fa067c6a7a/nack' \
  -H 'accept: application/json'

HTTP/1.1 204 No Content
date: Thu, 06 Oct 2022 01:10:26 GMT
server: uvicorn
content-type: application/json
```

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/all-events/messages' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Thu, 06 Oct 2022 01:10:49 GMT
server: uvicorn
content-length: 700
content-type: application/json

{
  "data":[
    {
      "id":"1dd422ad-7d20-4257-94c0-00fc1fbb8092",
      "queue_id":"all-events",
      "data":{
        "success":true,
        "event_name":"event1"
      },
      "attributes":{
        "event_name":"event1"
      },
      "delivery_attempts":2,
      "expired_at":"2022-10-19T23:43:33.265369",
      "scheduled_at":"2022-10-06T01:11:19.631164",
      "created_at":"2022-10-05T23:43:33.265369",
      "updated_at":"2022-10-06T01:10:49.631164"
    },
    {
      "id":"8856ea21-665f-40a1-b576-33fa067c6a7a",
      "queue_id":"all-events",
      "data":{
        "success":true,
        "event_name":"event2"
      },
      "attributes":{
        "event_name":"event2"
      },
      "delivery_attempts":2,
      "expired_at":"2022-10-19T23:51:36.175915",
      "scheduled_at":"2022-10-06T01:11:19.631164",
      "created_at":"2022-10-05T23:51:36.175915",
      "updated_at":"2022-10-06T01:10:49.631164"
    }
  ]
}
```

```bash
curl -i -X 'PUT' \
  'http://localhost:8000/messages/1dd422ad-7d20-4257-94c0-00fc1fbb8092/nack' \
  -H 'accept: application/json'

HTTP/1.1 204 No Content
date: Thu, 06 Oct 2022 01:11:34 GMT
server: uvicorn
content-type: application/json
```

```bash
curl -i -X 'PUT' \
  'http://localhost:8000/messages/8856ea21-665f-40a1-b576-33fa067c6a7a/nack' \
  -H 'accept: application/json'

HTTP/1.1 204 No Content
date: Thu, 06 Oct 2022 01:12:03 GMT
server: uvicorn
content-type: application/json
```

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/all-events/stats' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Thu, 06 Oct 2022 01:13:14 GMT
server: uvicorn
content-length: 69
content-type: application/json

{"num_undelivered_messages":0,"oldest_unacked_message_age_seconds":0}
```

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/all-events-dead/stats' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Thu, 06 Oct 2022 01:13:43 GMT
server: uvicorn
content-length: 72
content-type: application/json

{"num_undelivered_messages":2,"oldest_unacked_message_age_seconds":5410}
```

## Redrive support

With redrive you can move messages between queues, we usually use this feature to move messages from the dead queue to the original queue.

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/all-events/stats' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Fri, 07 Oct 2022 20:13:15 GMT
server: uvicorn
content-length: 69
content-type: application/json

{"num_undelivered_messages":0,"oldest_unacked_message_age_seconds":0}
```

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/all-events-dead/stats' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Fri, 07 Oct 2022 20:13:36 GMT
server: uvicorn
content-length: 71
content-type: application/json

{"num_undelivered_messages":2,"oldest_unacked_message_age_seconds":573}
```

```bash
curl -i -X 'PUT' \
  'http://localhost:8000/queues/all-events-dead/redrive' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "destination_queue_id": "all-events"
}'

HTTP/1.1 204 No Content
date: Fri, 07 Oct 2022 20:16:39 GMT
server: uvicorn
content-type: application/json
```

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/all-events/stats' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Fri, 07 Oct 2022 20:17:09 GMT
server: uvicorn
content-length: 71
content-type: application/json

{"num_undelivered_messages":2,"oldest_unacked_message_age_seconds":787}
```

## Delay queues

Delay queues let you postpone the delivery of new messages to consumers for a number of seconds.

```bash
curl -i -X 'POST' \
  'http://localhost:8000/queues' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "id": "delayed-events",
  "topic_id": "events",
  "ack_deadline_seconds": 30,
  "message_retention_seconds": 1209600,
  "delivery_delay_seconds": 30
}'

HTTP/1.1 201 Created
date: Sat, 19 Nov 2022 17:53:42 GMT
server: uvicorn
content-length: 291
content-type: application/json

{
  "id":"delayed-events",
  "topic_id":"events",
  "dead_queue_id":null,
  "ack_deadline_seconds":30,
  "message_retention_seconds":1209600,
  "message_filters":null,
  "message_max_deliveries":null,
  "delivery_delay_seconds":30,
  "created_at":"2022-11-19T17:53:42.858140",
  "updated_at":"2022-11-19T17:53:42.858140"
}
```

```bash
curl -i -X 'POST' \
  'http://localhost:8000/topics/events/messages' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "data": {"event_name": "event3", "success": true},
  "attributes": {"event_name": "event3"}
}'

HTTP/1.1 201 Created
date: Sat, 19 Nov 2022 18:02:08 GMT
server: uvicorn
content-length: 359
content-type: application/json

{
  "data":[
    {
      "id":"ed42463c-7725-4c3d-9de0-142bb0074be6",
      "queue_id":"delayed-events",
      "data":{
        "event_name":"event3",
        "success":true
      },
      "attributes":{
        "event_name":"event3"
      },
      "delivery_attempts":0,
      "expired_at":"2022-12-03T18:02:09.435449",
      "scheduled_at":"2022-11-19T18:02:39.435449",
      "created_at":"2022-11-19T18:02:09.435449",
      "updated_at":"2022-11-19T18:02:09.435449"
    }
  ]
}
```

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/delayed-events/messages' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Sat, 19 Nov 2022 18:02:24 GMT
server: uvicorn
content-length: 11
content-type: application/json

{"data":[]}
```

Wait for 30 seconds and try again:

```bash
curl -i -X 'GET' \
  'http://localhost:8000/queues/delayed-events/messages' \
  -H 'accept: application/json'

HTTP/1.1 200 OK
date: Sat, 19 Nov 2022 18:02:58 GMT
server: uvicorn
content-length: 359
content-type: application/json

{
  "data":[
    {
      "id":"ed42463c-7725-4c3d-9de0-142bb0074be6",
      "queue_id":"delayed-events",
      "data":{
        "success":true,
        "event_name":"event3"
      },
      "attributes":{
        "event_name":"event3"
      },
      "delivery_attempts":1,
      "expired_at":"2022-12-03T18:02:09.435449",
      "scheduled_at":"2022-11-19T18:03:28.853622",
      "created_at":"2022-11-19T18:02:09.435449",
      "updated_at":"2022-11-19T18:02:58.853622"
    }
  ]
}
```

## Prometheus metrics

You can enable prometheus metrics using the environment variable `fastqueue_enable_prometheus_metrics='true'`.

```bash
docker run --name fastqueue-server \
    --restart unless-stopped \
    -e fastqueue_database_url='postgresql+psycopg2://fastqueue:fastqueue@localhost:5432/fastqueue' \
    -e fastqueue_server_port=8000 \
    -e fastqueue_server_num_workers=1 \
    -e fastqueue_enable_prometheus_metrics='true' \
    --network="host" \
    quay.io/allisson/fastqueue server
```

```bash
curl -i -X 'GET' 'http://localhost:8000/metrics'

HTTP/1.1 200 OK
date: Fri, 14 Oct 2022 15:58:56 GMT
server: uvicorn
content-length: 4432
content-type: text/plain; version=0.0.4; charset=utf-8

# HELP python_gc_objects_collected_total Objects collected during gc
# TYPE python_gc_objects_collected_total counter
python_gc_objects_collected_total{generation="0"} 427.0
python_gc_objects_collected_total{generation="1"} 198.0
python_gc_objects_collected_total{generation="2"} 16.0
# HELP python_gc_objects_uncollectable_total Uncollectable object found during GC
# TYPE python_gc_objects_uncollectable_total counter
python_gc_objects_uncollectable_total{generation="0"} 0.0
python_gc_objects_uncollectable_total{generation="1"} 0.0
python_gc_objects_uncollectable_total{generation="2"} 0.0
# HELP python_gc_collections_total Number of times this generation was collected

CHANGELOG.md
# Changelog

## [1.7.0](https://github.com/allisson/fastqueue/compare/v1.6.0...v1.7.0) (2023-04-26)


### Features

* add an endpoint for health check ([#50](https://github.com/allisson/fastqueue/issues/50)) ([fd102bc](https://github.com/allisson/fastqueue/commit/fd102bc3bd2bd68d01cd95d4479f67b654ce61bd))
* upgrade sqlalchemy to version 2.x ([#48](https://github.com/allisson/fastqueue/issues/48)) ([2f74302](https://github.com/allisson/fastqueue/commit/2f74302d683068a2dda347d35d356b68d2d26a6c))

## [1.6.0](https://github.com/allisson/fastqueue/compare/v1.5.0...v1.6.0) (2023-03-16)


### Features

* pin dependencies on major version ([#46](https://github.com/allisson/fastqueue/issues/46)) ([289baf4](https://github.com/allisson/fastqueue/commit/289baf4a3d0a921564114b4632d8ffcc6e96b94e))

## [1.5.0](https://github.com/allisson/fastqueue/compare/v1.4.0...v1.5.0) (2023-01-11)


### Features

* convert services to be used as instances ([#44](https://github.com/allisson/fastqueue/issues/44)) ([8e95aea](https://github.com/allisson/fastqueue/commit/8e95aea14f924ac3d46f37ea8a55c461e60d7126))

## [1.4.0](https://github.com/allisson/fastqueue/compare/v1.3.2...v1.4.0) (2022-11-19)


### Features

* add delay queues support ([#41](https://github.com/allisson/fastqueue/issues/41)) ([62b5db6](https://github.com/allisson/fastqueue/commit/62b5db6f60d3bfc0c2baee81903883bdebab80ce))

## [1.3.2](https://github.com/allisson/fastqueue/compare/v1.3.1...v1.3.2) (2022-10-25)


### Bug Fixes

* revert to psycopg2-binary ([#39](https://github.com/allisson/fastqueue/issues/39)) ([662b199](https://github.com/allisson/fastqueue/commit/662b199d17da691ee44d320007bac78841cfd175))

## [1.3.1](https://github.com/allisson/fastqueue/compare/v1.3.0...v1.3.1) (2022-10-25)


### Bug Fixes

* replace psycopg2-binary with psycopg2 ([#37](https://github.com/allisson/fastqueue/issues/37)) ([056323f](https://github.com/allisson/fastqueue/commit/056323f54d510148bd02b206e2604743b1837b91))

## [1.3.0](https://github.com/allisson/fastqueue/compare/v1.2.0...v1.3.0) (2022-10-25)


### Features

* use python 3.11 ([#35](https://github.com/allisson/fastqueue/issues/35)) ([162078d](https://github.com/allisson/fastqueue/commit/162078d8eb1f6ce7b9f63197ef432ca077399162))

## [1.2.0](https://github.com/allisson/fastqueue/compare/v1.1.0...v1.2.0) (2022-10-14)


### Features

* add prometheus metrics support ([#33](https://github.com/allisson/fastqueue/issues/33)) ([a624be1](https://github.com/allisson/fastqueue/commit/a624be14b84b88120239ee13a3fd518d048b749e))

## [1.1.0](https://github.com/allisson/fastqueue/compare/v1.0.0...v1.1.0) (2022-10-07)


### Features

* add redrive support ([#31](https://github.com/allisson/fastqueue/issues/31)) ([3eac767](https://github.com/allisson/fastqueue/commit/3eac7678b883df409eb3dbcc9bd63f7ac82e175b))

## 1.0.0 (2022-10-06)


### Features

* add ack and nack endpoints ([#13](https://github.com/allisson/fastqueue/issues/13)) ([940a5c6](https://github.com/allisson/fastqueue/commit/940a5c698136d2ff94f2f0735e113d01a5657378))
* add base models ([#4](https://github.com/allisson/fastqueue/issues/4)) ([f47cf1c](https://github.com/allisson/fastqueue/commit/f47cf1cea200eb54a6c631f8bcf1c1738942d2e1))
* add basic fastapi setup ([#2](https://github.com/allisson/fastqueue/issues/2)) ([12e71a5](https://github.com/allisson/fastqueue/commit/12e71a5c730d6cc64d842ef10c5b7a5d7f9e06b7))
* add database echo support ([#26](https://github.com/allisson/fastqueue/issues/26)) ([15b383f](https://github.com/allisson/fastqueue/commit/15b383fb68c8c350ab2f08e4264c5fafad484cbd))
* add database migrations with alembic ([#3](https://github.com/allisson/fastqueue/issues/3)) ([4bb4ae5](https://github.com/allisson/fastqueue/commit/4bb4ae5376b2af521e72cdd3e6cf7d1fc2a97dd6))
* add dead queue support ([#18](https://github.com/allisson/fastqueue/issues/18)) ([8e32aa0](https://github.com/allisson/fastqueue/commit/8e32aa0cd78d9a55d271c0af8e4e17cd3e77e7d2))
* add endpoint to purge the queue messages ([#22](https://github.com/allisson/fastqueue/issues/22)) ([b55c7fd](https://github.com/allisson/fastqueue/commit/b55c7fdca1efeef3ea702991d7a7782b1944c755))
* add MessageService ([#9](https://github.com/allisson/fastqueue/issues/9)) ([bf328bf](https://github.com/allisson/fastqueue/commit/bf328bfa80a86b9ff6730facf50631858ccd6105))
* add MessageService.list_for_consume ([#10](https://github.com/allisson/fastqueue/issues/10)) ([901bd1b](https://github.com/allisson/fastqueue/commit/901bd1b443b7008d273ea2fcfc4aed90746584d7))
* add queue stats endpoint ([#14](https://github.com/allisson/fastqueue/issues/14)) ([76c906a](https://github.com/allisson/fastqueue/commit/76c906a3c8ef3afeff2a57672ae8ec40c63aa487))
* add queues and messages endpoints ([#12](https://github.com/allisson/fastqueue/issues/12)) ([be80f17](https://github.com/allisson/fastqueue/commit/be80f17327cc4c9e318601a442f02d8bf9083fce))
* add QueueService ([#8](https://github.com/allisson/fastqueue/issues/8)) ([29c9768](https://github.com/allisson/fastqueue/commit/29c9768ceb87b5e59ee7f6fda5ab51373973fee5))
* add QueueService cleanup method ([#16](https://github.com/allisson/fastqueue/issues/16)) ([f15f30a](https://github.com/allisson/fastqueue/commit/f15f30a41ea33fa737addcbd522d88fea0954c73))
* add topics endpoints ([#11](https://github.com/allisson/fastqueue/issues/11)) ([f8dd58a](https://github.com/allisson/fastqueue/commit/f8dd58a21b8b3add74686cc54850b82778d17b40))
* add TopicService ([#7](https://github.com/allisson/fastqueue/issues/7)) ([08ccad7](https://github.com/allisson/fastqueue/commit/08ccad7b3adc8a4e506177ea83ab775dc38aec48))
* add worker entrypoint ([#17](https://github.com/allisson/fastqueue/issues/17)) ([6627950](https://github.com/allisson/fastqueue/commit/662795060da39e0b15f1bd8fbe188d7c6c43a82f))
* configure the base setup on github ([#1](https://github.com/allisson/fastqueue/issues/1)) ([803b70c](https://github.com/allisson/fastqueue/commit/803b70c78c7fa4b0e4fda2576c0b660d8ddf748d))

alembic/versions/001_initial.py
"""Auto generated

Revision ID: fe94d60449f2
Revises:
Create Date: 2022-09-15 17:51:40.259256

"""
import sqlalchemy as sa
from sqlalchemy.dialects import postgresql

from alembic import op

# revision identifiers, used by Alembic.
revision = "fe94d60449f2"
down_revision = None
branch_labels = None
depends_on = None


def upgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table(
        "topics",
        sa.Column("id", sa.String(length=128), nullable=False),
        sa.Column("created_at", sa.DateTime(), nullable=False),
        sa.PrimaryKeyConstraint("id"),
    )
    op.create_table(
        "queues",
        sa.Column("id", sa.String(length=128), nullable=False),
        sa.Column("topic_id", sa.String(length=128), nullable=True),
        sa.Column("dead_queue_id", sa.String(length=128), nullable=True),
        sa.Column("ack_deadline_seconds", sa.Integer(), nullable=False),
        sa.Column("message_retention_seconds", sa.Integer(), nullable=False),
        sa.Column("message_filters", postgresql.JSONB(astext_type=sa.Text()), nullable=True),
        sa.Column("message_max_deliveries", sa.Integer(), nullable=True),
        sa.Column("created_at", sa.DateTime(), nullable=False),
        sa.Column("updated_at", sa.DateTime(), nullable=False),
        sa.ForeignKeyConstraint(
            ["dead_queue_id"],
            ["queues.id"],
        ),
        sa.ForeignKeyConstraint(
            ["topic_id"],
            ["topics.id"],
        ),
        sa.PrimaryKeyConstraint("id"),
    )
    op.create_index(op.f("ix_queues_topic_id"), "queues", ["topic_id"], unique=False)
    op.create_table(
        "messages",
        sa.Column("id", postgresql.UUID(), nullable=False),
        sa.Column("queue_id", sa.String(length=128), nullable=False),
        sa.Column("data", postgresql.JSONB(astext_type=sa.Text()), nullable=False),
        sa.Column("attributes", postgresql.JSONB(astext_type=sa.Text()), nullable=True),
        sa.Column("delivery_attempts", sa.Integer(), nullable=False),
        sa.Column("expired_at", sa.DateTime(), nullable=False),
        sa.Column("scheduled_at", sa.DateTime(), nullable=False),
        sa.Column("created_at", sa.DateTime(), nullable=False),
        sa.Column("updated_at", sa.DateTime(), nullable=False),
        sa.ForeignKeyConstraint(
            ["queue_id"],
            ["queues.id"],
        ),
        sa.PrimaryKeyConstraint("id"),
    )
    op.create_index(
        "ix_messages_expired_at", "messages", ["expired_at"], unique=False, postgresql_using="brin"
    )
    op.create_index(
        "ix_messages_expired_at_scheduled_at",
        "messages",
        ["expired_at", "scheduled_at"],
        unique=False,
        postgresql_using="brin",
    )
    op.create_index(
        "ix_messages_scheduled_at", "messages", ["scheduled_at"], unique=False, postgresql_using="brin"
    )
    # ### end Alembic commands ###


def downgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    op.drop_index("ix_messages_scheduled_at", table_name="messages", postgresql_using="brin")
    op.drop_index("ix_messages_expired_at_scheduled_at", table_name="messages", postgresql_using="brin")
    op.drop_index("ix_messages_expired_at", table_name="messages", postgresql_using="brin")
    op.drop_table("messages")
    op.drop_index(op.f("ix_queues_topic_id"), table_name="queues")
    op.drop_table("queues")
    op.drop_table("topics")
    # ### end Alembic commands ###
